From dev to main (week #4 & 5)

ZhosanDenis opened the pull request.

ZhosanDenis committed:
Изменения №1 от 24.06.2023
- добавлен класс-обертка AppUser, реализующий UserDetails и хранящий UserEntity;
- добавлен FromDbUserManager, реализующий UserDetailsManager и использующийся AuthServiceImpl и AccountServiceImpl для операций с UserEntity и БД;
- добавлена функциональность по созданию пользователя с правами ADMIN при запуске приложения.

ZhosanDenis committed:
Изменения №1 от 25.06.2023
- добавлен функционал по изменению/удалению объявления или комментария со стороны создавшего пользователя или админа

ZhosanDenis committed:
Изменения №2 от 25.06.2023
- внесены изменения, связанные с сохранением в классе AppUser ДТО сущности, а не самой сущности.

ZhosanDenis committed:
Merge pull request #16 from Zhoss/1-feature-denis

Изменения №1 от 24.06.2023

dmitry-bizin commented on src/main/java/ru/skypro/homework/WebSecurityConfig.java:
```diff
@@ -25,15 +24,8 @@ public class WebSecurityConfig {
-            .roles("USER")
-            .build();
-    return new InMemoryUserDetailsManager(user);
+  public FromDbUserManager userDetailsService() {
```
здесь лучше бы возвращаемый тип сделать UserDetailsManager

dmitry-bizin commented on src/main/java/ru/skypro/homework/security/FromDbUserManager.java:
```diff
@@ -0,0 +1,63 @@
+
+public class FromDbUserManager implements UserDetailsManager {
+
+    @Autowired
```
а почему не через конструктор? лучше бы через конструктор инъекции делать

dmitry-bizin replied:
и сделать тогда этот класс бином, т.е. с аннотацией Service, а из WebSecurityConfig удалить метод с аннотацией Bean

dmitry-bizin commented on src/main/java/ru/skypro/homework/security/FromDbUserManager.java:
```diff
@@ -0,0 +1,63 @@
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        UserEntity userEntity = userRepository.findByEmail(username)
+                .orElseThrow(() -> new IllegalArgumentException("Пользователь не найден"));
```
нужно кинуть UsernameNotFoundException

dmitry-bizin commented on src/main/java/ru/skypro/homework/security/FromDbUserManager.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Override
+    public void changePassword(String oldPassword, String newPassword) {
+        Authentication currentUser = SecurityContextHolder.getContext().getAuthentication();
```
в целом, т.к. смена пароля под аутентификацией, то юзер уже будет установлен в AppUser, поэтому можно было бы его использовать (если бы он был бином, как я показывал на разборе)

dmitry-bizin commented on src/main/java/ru/skypro/homework/security/FromDbUserManager.java:
```diff
@@ -0,0 +1,63 @@
+        Authentication currentUser = SecurityContextHolder.getContext().getAuthentication();
+        String userName = currentUser.getName();
+        UserEntity userEntity = userRepository.findByEmail(userName)
+                .orElseThrow(() -> new IllegalArgumentException("Пользователь не найден"));
```
лучше кинуть UsernameNotFoundException

dmitry-bizin commented on src/main/java/ru/skypro/homework/security/FromDbUserManager.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Override
+    public void createUser(UserDetails user) {
+        UserDetails userDetails = loadUserByUsername(user.getUsername());
```
а зачем доставать из БД, если он уже здесь есть в параметре?)

dmitry-bizin replied:
этот метод не получится использовать при регистрации, т.к. тогда придётся данные регистрации оборачивать в UserDetails, а это не совсем то, что нужно

dmitry-bizin commented on src/main/java/ru/skypro/homework/security/FromDbUserManager.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Override
+    public void updateUser(UserDetails user) {
+        UserDetails userDetails = loadUserByUsername(user.getUsername());
```
а зачем доставать из БД, если он уже здесь есть в параметре?)

dmitry-bizin replied:
этот метод не получится использовать при обновлении юзера, т.к. тогда придётся данные обновления оборачивать в UserDetails, а это не совсем то, что нужно

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/AccountServiceImpl.java:
```diff
@@ -25,65 +29,72 @@
-                )
-        );
+        UserEntity updatedUser = userRepository.save(userMapper.updateUserEntity(userEntity, user));
+        manager.updateUser(new AppUser(securityUserMapper.toUserDto(updatedUser)));
```
хм, но ведь данные не обновятся. Потому что в UserDto у тебя нет ни фамилии, ни имени, ни телефона. Здесь не нужно использовать manager, надо обновлять без него, как я и говорил на разборе

dmitry-bizin replied:
а, ты выше обновляешь же) короче, эта строка не нужна

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/AccountServiceImpl.java:
```diff
@@ -25,65 +29,72 @@
+        userEntity.setImageMediaType(image.getContentType());
+        userEntity.setImageFileSize(image.getSize());
+        ((AppUser) user).setUser(securityUserMapper.toUserDto(userEntity));
+        manager.updateUser(user);
```
не нужна эта строка

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/AdServiceImpl.java:
```diff
@@ -184,4 +225,12 @@ private Path createPath(MultipartFile image, AdEntity adEntity) throws IOExcepti
     }
+
+    private boolean userCanChangeAdvertising(UserEntity user, AdEntity ad) {
+        return user.getRole().equals(Role.ADMIN) || ad.getUserEntity().equals(user);
```
enum-ы можно через == сравнивать

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/AdServiceImpl.java:
```diff
@@ -184,4 +225,12 @@ private Path createPath(MultipartFile image, AdEntity adEntity) throws IOExcepti
+    }
+
+    private boolean userCanChangeComment(UserEntity user, CommentEntity comment) {
+        return user.getRole().equals(Role.ADMIN) || comment.getUserEntity().equals(user);
```
enum-ы можно через == сравнивать

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/AuthServiceImpl.java:
```diff
@@ -23,42 +27,62 @@ public class AuthServiceImpl implements AuthService {
-                        .roles(role.name())
-                        .build());
-        userRepository.save(mapper.toUserEntity(registerReq));
+        manager.createUser(new AppUser(
```
не нужно manager вызывать

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/AdController.java:
```diff
@@ -125,7 +126,11 @@ public ResponseEntity<ResponseWrapperAds> getAllMyAdvertising(Authentication aut
+        if (updatedAds != null) {
+            return ResponseEntity.ok(updatedAds);
+        }
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
```
forbidden

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/AdController.java:
```diff
@@ -161,8 +166,10 @@ public ResponseEntity<?> updateAdvertisingImage(@PathVariable int id,
+        if (adService.deleteAdvertising(id)) {
+            return ResponseEntity.ok().build();
+        }
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
```
forbidden

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/AdController.java:
```diff
@@ -225,7 +232,11 @@ public ResponseEntity<Comment> updateComment(
+        if (comment != null) {
+            return ResponseEntity.ok(comment);
+        }
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
```
forbidden

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/AdController.java:
```diff
@@ -241,8 +252,10 @@ public ResponseEntity<Comment> updateComment(
+        if (adService.deleteComment(adId, commentId)) {
+            return ResponseEntity.ok().build();
+        }
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
```
forbidden

ZhosanDenis committed:
Изменения №1 от 27.06.2023
Исправлены замечания от наставника, связанные с использованием UserDetails и UserDetailsManager. Теперь класс, реализующий UserDetails является бином, создаваемым при запросе. Информация об аутентифицированном пользователе берется из данного бина, а не из объекта Authentication, как ранее.

ZhosanDenis committed:
Изменения №1 от 29.06.2023
- внесены изменения, предотвращающие выброс NPE в запросах по получению информации об объявлении;
- исправлены ошибки, из-за которых не загружались изображения на фронт.

ZhosanDenis committed:
Изменения №1 от 03.07.2023
- внесены изменения, влияющие на корректное отображение аватарки пользователя в комментариях к объявлению;
- в конфигурации безопасности добавлен разрешение для эндпоинта, загружающего аватарку пользователя - данный эндпоинт теперь доступен для любого пользователя, в том числе не аутентифицированного.

ZhosanDenis committed:
Merge pull request #18 from Zhoss/1-feature-denis

Изменения №1 от 27.06.2023

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/AccountServiceImpl.java:
```diff
@@ -40,17 +39,17 @@ public class AccountServiceImpl implements AccountService {
-    public boolean updatePassword(NewPassword newPassword, String userName) {
-        if (userRepository.existsByEmail(userName)) {
+    public boolean updatePassword(NewPassword newPassword) {
+        if (userDetails != null) {
```
эта проверка точно нужна? он же не будет никогда null

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/AdServiceImpl.java:
```diff
@@ -187,10 +187,9 @@ public boolean updateAdvertisingImage(int id, MultipartFile image) throws IOExce
+                .orElseThrow(() -> new UsernameNotFoundException("Пользователь не найден"));
         AdEntity adEntity = adRepository.findById(id)
                 .orElseThrow(() -> new IllegalArgumentException("Объявление не найдено"));
         if (userCanChangeAdvertising(userEntity, adEntity)) {
```
картинку из файловой системы тоже надо удалять

ZhosanDenis committed:
Изменения №1 от 06.07.2023
- внесены изменения, связанные с удалением из файловой системы изображений пользователя (при обновлении аватарки - если изображения разных форматов, то у одного пользователя было несколько изображений) и изображений объявлений (при удалении объявления или при обновлении изображений)

ZhosanDenis committed:
Merge pull request #19 from Zhoss/1-feature-denis

Изменения №1 от 06.07.2023
